atoma-proxy/src/telemetry.rs

```diff
@@ -0,0 +1,195 @@
+use anyhow::{Context, Result};
+use once_cell::sync::Lazy;
+use opentelemetry::{
+    global,
+    metrics::{Counter, Histogram, Meter, Unit},
+    trace::TracerProvider,
+    KeyValue,
+};
+use opentelemetry_otlp::{new_exporter, WithExportConfig};
+use opentelemetry_sdk::{
+    metrics::{self as sdkmetrics},
+    trace as sdktrace, Resource,
+};
+use std::path::Path;
+use tracing_appender::non_blocking::WorkerGuard;
+use tracing_appender::{
+    non_blocking,
+    rolling::{RollingFileAppender, Rotation},
+};
+use tracing_opentelemetry::OpenTelemetryLayer;
+use tracing_subscriber::{
+    fmt::{self, format::FmtSpan, time::UtcTime},
+    layer::SubscriberExt,
+    util::SubscriberInitExt,
+    EnvFilter, Registry,
+};
+
+const LOG_FILE: &str = "atoma-proxy-service.log";
+// Default Grafana OTLP endpoint if not specified in environment
+const DEFAULT_OTLP_ENDPOINT: &str = "http://localhost:4317";
+
+static RESOURCE: Lazy<Resource> =
+    Lazy::new(|| Resource::new(vec![KeyValue::new("service.name", "atoma-proxy")]));
+
+// Add global metrics
+static GLOBAL_METER: Lazy<Meter> = Lazy::new(|| global::meter("atoma-proxy"));
+
+// Define metric instruments
+static REQUEST_DURATION: Lazy<Histogram<f64>> = Lazy::new(|| {
+    GLOBAL_METER
+        .f64_histogram("request_duration")
+        .with_description("Total request latency in seconds")
+        .with_unit(Unit::new("s"))
+        .init()
+});
+
+static MIDDLEWARE_DURATION: Lazy<Histogram<f64>> = Lazy::new(|| {
+    GLOBAL_METER
+        .f64_histogram("middleware_duration")
+        .with_description("Middleware execution time in seconds")
+        .with_unit(Unit::new("s"))
+        .init()
+});
+
+static DB_READ_DURATION: Lazy<Histogram<f64>> = Lazy::new(|| {
+    GLOBAL_METER
+        .f64_histogram("db_read_duration")
+        .with_description("Database read operation duration in seconds")
+        .with_unit(Unit::new("s"))
+        .init()
+});
+
+static DB_WRITE_DURATION: Lazy<Histogram<f64>> = Lazy::new(|| {
+    GLOBAL_METER
+        .f64_histogram("db_write_duration")
+        .with_description("Database write operation duration in seconds")
+        .with_unit(Unit::new("s"))
+        .init()
+});
+
+static DB_READ_COUNT: Lazy<Counter<u64>> = Lazy::new(|| {
+    GLOBAL_METER
+        .u64_counter("db_read_count")
+        .with_description("Number of database read operations")
+        .init()
+});
+
+static DB_WRITE_COUNT: Lazy<Counter<u64>> = Lazy::new(|| {
+    GLOBAL_METER
+        .u64_counter("db_write_count")
+        .with_description("Number of database write operations")
+        .init()
+});
+
+/// Initialize metrics with OpenTelemetry SDK
+fn init_metrics() -> sdkmetrics::MeterProvider {
+    sdkmetrics::MeterProvider::builder()
+        .with_resource(RESOURCE.clone())
+        .build()
+}
+
+/// Initialize tracing with OpenTelemetry SDK
+fn init_traces() -> Result<sdktrace::Tracer> {
```

Comment:
Maybe good to have proper instrumentation and logging

Reply:
What do you mean by this @jorgeantonio21 ?